File: amazon-scraper/node.js/singleProduct.js
const axios = require('axios');
const cheerio = require('cheerio');
const createCsvWriter = require('csv-writer').createObjectCsvWriter;

// Configuration
const token = "<SDO-token>";
const asins = ["B0BLRJ4R8F", "B081YXWDTQ", "B07D74DT3B"]; // Change these to any product ASINs
const geocode = "us";
const zipcode = "10001";

async function scrapeSingleProducts() {
    const allProducts = [];

    // Loop through all ASINs
    for (const asin of asins) {
        // Build product URL from ASIN
        const targetUrl = `https://www.amazon.com/dp/${asin}`;

        // Make API request
        const encodedUrl = encodeURIComponent(targetUrl);
        const apiUrl = `https://api.scrape.do/plugin/amazon/?token=${token}&url=${encodedUrl}&geocode=${geocode}&zipcode=${zipcode}&output=html`;

        try {
            const response = await axios.get(apiUrl);
            const $ = cheerio.load(response.data);

            // Extract product details
            const nameElem = $('#productTitle');
            const name = nameElem.text().trim();

            if (!name) {
                console.log(`Error scraping ${asin}: Product title not found`);
                continue;
            }

            // Extract price
            let price;
            if ($('#outOfStockBuyBox_feature_div').length > 0) {
                price = 'Out of Stock';
            } else {
                const whole = $('.a-price-whole').first().text();
                const fraction = $('.a-price-fraction').first().text();
                price = whole && fraction ? `$${whole}${fraction}` : 'N/A';
            }

            // Extract image
            const imageElem = $('#landingImage');
            const image = imageElem.attr('src') || 'N/A';

            // Extract rating
            const ratingElem = $('.AverageCustomerReviews');
            const ratingText = ratingElem.text().trim();
            const rating = ratingText ? ratingText.split(' out of')[0] : 'N/A';

            allProducts.push({
                ASIN: asin,
                Name: name,
                Price: price,
                Image: image,
                Rating: rating
            });

            console.log(`Scraped: ${asin} - ${name.substring(0, 50)}...`);
        } catch (e) {
            console.log(`Error scraping ${asin}: ${e.message}`);
            continue;
        }
    }

    // Export to CSV
    const csvWriter = createCsvWriter({
        path: 'singleProduct.csv',
        header: [
            { id: 'ASIN', title: 'ASIN' },
            { id: 'Name', title: 'Name' },
            { id: 'Price', title: 'Price' },
            { id: 'Image', title: 'Image' },
            { id: 'Rating', title: 'Rating' }
        ]
    });

    await csvWriter.writeRecords(allProducts);

    console.log(`\nScraped ${allProducts.length} products`);
    console.log('Data exported to singleProduct.csv');
}

scrapeSingleProducts().catch(console.error);

File: amazon-scraper/python/singleProduct.py
import urllib.parse
import csv
import requests
from bs4 import BeautifulSoup

# Configuration
token = "<SDO-token>"
asins = ["B0BLRJ4R8F", "B081YXWDTQ", "B07D74DT3B"]  # Change these to any product ASINs
geocode = "us"
zipcode = "10001"

all_products = []

# Loop through all ASINs
for asin in asins:
    # Build product URL from ASIN
    target_url = f"https://www.amazon.com/dp/{asin}"

    # Make API request
    targetUrl = urllib.parse.quote(target_url)
    apiUrl = "https://api.scrape.do/plugin/amazon/?token={}&url={}&geocode={}&zipcode={}&output=html".format(token, targetUrl, geocode, zipcode)
    response = requests.request("GET", apiUrl)

    soup = BeautifulSoup(response.text, "html.parser")

    # Extract product details
    try:
        name = soup.find(id="productTitle").text.strip()

        # Extract price
        if soup.find("div", id="outOfStockBuyBox_feature_div"):
            price = "Out of Stock"
        else:
            whole = soup.find(class_="a-price-whole")
            fraction = soup.find(class_="a-price-fraction")
            price = f"${whole.text}{fraction.text}" if whole and fraction else "N/A"

        # Extract image
        image_elem = soup.find("img", {"id": "landingImage"})
        image = image_elem["src"] if image_elem else "N/A"

        # Extract rating
        rating_elem = soup.find(class_="AverageCustomerReviews")
        rating = rating_elem.text.strip().split(" out of")[0] if rating_elem else "N/A"

        all_products.append({
            "ASIN": asin,
            "Name": name,
            "Price": price,
            "Image": image,
            "Rating": rating
        })

        print(f"Scraped: {asin} - {name[:50]}...")
    except Exception as e:
        print(f"Error scraping {asin}: {e}")
        continue

# Export to CSV
csv_file = "singleProduct.csv"
headers = ["ASIN", "Name", "Price", "Image", "Rating"]

with open(csv_file, "w", newline="", encoding="utf-8") as f:
    writer = csv.DictWriter(f, fieldnames=headers)
    writer.writeheader()
    writer.writerows(all_products)

print(f"\nScraped {len(all_products)} products")
print(f"Data exported to {csv_file}")
